File: src/python/oop/TicTacToe/game/score_board.py
from .score_record import ScoreRecord
from .player import Player


class ScoreBoard:
    def __init__(self):
        self._score = {}

    def record_win(self, player: Player):
        if player.name not in self._score:
            self._score[player.name] = 0

        self._score[player.name] += 1

    def record_loose(self, player: Player):
        if player.name not in self._score:
            self._score[player.name] = 0

        self._score[player.name] -= 1

    def get_player_score(self, player: Player) -> ScoreRecord:
        return ScoreRecord(player.name, self._score[player.name] if player.name in self._score else 0)

    def get_top_players(self) -> list[ScoreRecord]:
        top = []
        for player, score in self._score.items():
            top.append(ScoreRecord(player, score))

        return sorted(top, key=lambda x: -x.score)

File: src/csharp/oop/TicTacToe/Game/ScoreBoard.cs
namespace TicTacToe.Game;

public class ScoreBoard
{
    private readonly Dictionary<string, int> _scores = new();

    public void RecordWin(Player player)
    {
        _scores.TryAdd(player.Name, 0);
        _scores[player.Name] += 1;
    }

    public void RecordLoose(Player player)
    {
        _scores.TryAdd(player.Name, 0);
        _scores[player.Name] -= 1;
    }

    public ScoreRecord GetPlayerScore(Player player)
        => new (player.Name, _scores.GetValueOrDefault(player.Name, 0));

    public IReadOnlyCollection<ScoreRecord> GetScores()
        => _scores
            .Select(it => new ScoreRecord(it.Key, it.Value))
            .OrderByDescending(it => it.Score)
            .ToList();
}
